### ISSUE TITLE
PubSub SubscriberServiceApiClient.PullAsync doesn't respect maxMessages

### ISSUE THREAD

**Issue by Rassi** [googleapis/google-cloud-dotnet#7551]
TLDR: Using SubscriberServiceApiClient.PullAsync with maxMessages set to 100, many times return (significantly) less than 100, even though the subscription has many thousands of messages.

### Issue
We are switching from a different message queue system to PubSub, and have external partners consuming messages from our subscriptions. One of those partners consumes messages in a cron-job-ish way, where they start a process which polls for messages every minute or so. They monitor the size and age of their subscription, and are experiencing situations where they aren't handling messages in a timely manner.

We have identified the issue being the fact that they expected to receive 100 messages at a time, using `SubscriberServiceApiClient.PullAsync` since they specified maxMessages at 100. Testing it in a dotnet 6 console app I reproduced it easily, pulling from a subscription with more than 1000 messages:

```cs
using Google.Cloud.PubSub.V1; // 2.7.0

var subscriptionName = SubscriptionName.FromProjectSubscription(projectId, subscriptionId);
var subscriberClient = SubscriberServiceApiClient.Create();

for (int i = 0; i < 10; i++)
{
    var pullResponse = await subscriberClient.PullAsync(subscriptionName, 100);
    Console.WriteLine($"PullAsync asked for {maxMessages} got {pullResponse.ReceivedMessages.Count}");
}
```
Producing the output:
```
PullAsync asked for 100 got 100
PullAsync asked for 100 got 100
PullAsync asked for 100 got 26
PullAsync asked for 100 got 17
PullAsync asked for 100 got 25
PullAsync asked for 100 got 38
PullAsync asked for 100 got 56
PullAsync asked for 100 got 84
PullAsync asked for 100 got 100
PullAsync asked for 100 got 24
```

I am aware that the [documentation ](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.PubSub.V1/latest/Google.Cloud.PubSub.V1.SubscriberServiceApiClient#Google_Cloud_PubSub_V1_SubscriberServiceApiClient_PullAsync_Google_Cloud_PubSub_V1_SubscriptionName_System_Int32_Google_Api_Gax_Grpc_CallSettings_) states that _"The Pub/Sub system may return fewer than the number specified."_. But I (and our partner) understood that as: if there are fewer messages in the subscription than maxMessages, then the amount that is in the subscription would be returned even though it is less than maxMessages (instead of e.g. blocking and waiting for more).

(Besides that, the documentation has examples which state `int maxMessages = 0`, even though the documentation says _"Must be a positive integer"_, but I digress)

Unless our code is wrong, I am assuming that the sentence in fact means that even though there are more messages in the subscription than maxMessages, we may still get fewer?

I am a big proponent of the Principle of Least Astonishment, and I must say that it is quite astonishing to call a pull method, and receive an arbitrary number of messages between 1 and maxMessages.

Is there a way to pull messages using `SubscriberServiceApiClient` and ensure that you receive the number of messages that you request?

**Comment by jskeet** [comment]
> Unless our code is wrong, I am assuming that the sentence in fact means that even though there are more messages in the subscription than maxMessages, we may still get fewer?

I believe so, but I'll need to check with the PubSub team.

**Comment by kamalaboulhosn** [comment]
The guarantee with pull is that you will receive up to `maxMessages`, but may receive fewer, even if there are more than `maxMessages` in the backlog. The reason is that the service is trading off fuller pull responses for shorter end-to-end latency. From the time a request comes in, messages are fetched to send in response. If some messages are quickly available, the choice is between waiting for more messages to become available or returning the messages that were quickly available. If the system waits, then the end-to-end latency increases for the messages that were already fetched. Therefore, Pub/Sub only waits a short period of time before sending the messages it has.

The code provided is an inefficient way to use Cloud Pub/Sub. The [synchronous pull](https://cloud.google.com/pubsub/docs/pull#synchronous_pull) section of the documentation says: 

>Note: To achieve low message delivery latency with synchronous pull, it is important to have many simultaneously outstanding pull requests. As the throughput of the topic increases, more pull requests are necessary. In general, asynchronous pull [via the client library using streaming pull] is preferable for latency-sensitive applications.

When there are a lot of simultaneous requests, the server knows to keep a constant flow of messages ready that can be returned in responses. When there is only one request outstanding at a time, the server doesn't know that there will necessarily be subsequent requests to fulfill and therefore doesn't keep a lot of messages waiting because we don't want to waste moving messages around.

There is no way to to ensure that you receive the number of messages you request with the Pull method in Cloud Pub/Sub.

**Comment by Rassi** [comment]
Thank you for the swift response. I still find it hard to not be able to point our clients in the direction of `SubscriberServiceApiClient` as a simple, deterministic and logical way to consume messages from our subscriptions, since we are not producing a high volume of messages, and don't need extremely high efficient throughput. Especially for clients working with polling cron-job style jobs running at fixed intervals.

I have tried looking at [Asynchronous Pull](https://cloud.google.com/pubsub/docs/pull#asynchronous-pull), which I suspect is what you're suggesting we use instead, and have gotten to the following point trying to get 200 messages.

```cs
var subscriptionName = SubscriptionName.FromProjectSubscription(projectId, subscriptionId);
var client = SubscriberClient.Create(subscriptionName);

var messagesProcessed = 0;
var messages = new ConcurrentBag<string>();
var messageProcessing = client.StartAsync((message, token) =>
{
    Interlocked.Increment(ref messagesProcessed);
    messages.Add(message.ToString());
    return Task.FromResult(SubscriberClient.Reply.Ack);
});

while (messagesProcessed < 200)
{
    await Task.Delay(1);
}

await client.StopAsync(TimeSpan.FromMilliseconds(500));
Console.WriteLine($"Done processing {messagesProcessed} messages. messages.count: {messages.Count}");
```

While this does give me at least 200 messages (and many times much more, even with a delay of only 1ms), this still gives me a nagging feeling that I might run into some of the issues that you describe in the [section about large backlogs](https://cloud.google.com/pubsub/docs/pull#streamingpull_dealing_with_large_backlogs_of_small_messages), such as having multiple nodes trying to get messages while a single node has many messages in it's backlog. It would still be much more preferable to ask for 200 messages simply, and only getting those and not affecting other messages which other nodes could process.

Is the code in the example above the way you would suggest we receive messages in batches of 200, with minimal influence on other nodes or the messages left in the subscription?

**Comment by Rassi** [comment]
Oh and there should probably be some kind of timeout so we don't block indefinitely if the subscription is empty. Or is there a simple way to get the subscription size so we can break out if it reaches 0?

**Comment by jskeet** [comment]
Again, these are really questions for the API team I'm afraid. The C# library for `SubscriberServiceApiClient` just exposes the underlying API. (`PublisherClient` and `SubscriberClient` are somewhat different, and requests for those classes *could* be in scope for this repository.)

I suggest that it's probably more productive to use one of the options on the [Pub/Sub support page](https://cloud.google.com/pubsub/docs/getting-support) - in particular, either posting a message in the [cloud-pubsub-discuss group](https://groups.google.com/forum/#!forum/cloud-pubsub-discuss) or filing a feature request in the [bug tracker](https://issuetracker.google.com/issues?q=componentid:187173%20status:open). Either of those will have more immediate visibility for the PubSub team than in this (C#-specific) repo.

I'll close this issue now as I don't believe there's anything we can do specifically in this repo. Obviously if you run into any C#-specific issues, please feel free to file another issue. Closing this issue isn't intended to imply that your comments/requests are unreasonable - just that the maintainers of this repository aren't able to help you with them.

**Comment by Rassi** [comment]
> PublisherClient and SubscriberClient are somewhat different, and requests for those classes could be in scope for this repository.

Please disregard if I'm off base, but would it then be possible to introduce a `PullAsync` method on `SubscriberClient` (like there's a [PublishAsync](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.PubSub.V1/latest/Google.Cloud.PubSub.V1.PublisherClient#Google_Cloud_PubSub_V1_PublisherClient_PublishAsync_Google_Cloud_PubSub_V1_PubsubMessage_) on `PublisherClient`) which would support a messageCount, that would ensure you get the amount of messages you ask for?

**Comment by jskeet** [comment]
> Please disregard if I'm off base, but would it then be possible to introduce a PullAsync method on SubscriberClient (like there's a PublishAsync on PublisherClient) which would support a messageCount, that would ensure you get the amount of messages you ask for?

That wouldn't really fit with the design of `SubscriberClient`, which is specifically a "start this and it will call you back" model. Even if we wanted to do that, we'd still end up with the exact same issue of the underlying API not supporting the feature you want - it's not like `SubscriberClient` has access to any "protected" API methods or anything :(
